### PR TITLE
Recurrent Batch Normalization from Tim Coojmans et al paper

### PR DESCRIPTION
**Pull request by lakehanne**

### CODE REVIEW

**Comment by nicholas-leonard** [comment]
@lakehanne  Your commits aren't registered to your github account. You can use ssh keys to link your local git client to your github account.

File: FastLSTM.lua [hunk 1]
```diff
@@ -3,22 +3,46 @@ local FastLSTM, parent = torch.class("nn.FastLSTM", "nn.LSTM")
...
-function FastLSTM:__init(inputSize, outputSize, rho)
-   parent.__init(self, inputSize, outputSize, rho, false)
+function FastLSTM:__init(inputSize, outputSize, rho, eps, momentum, affine)
+   --  initialize batch norm variance with 0.1
+   self.eps = self.eps or 0.1
+   self.momentum = self.momentum or 0.1 --gamma
+   self.affine = self.affine or true
+   self.bn = self.bn
+
+   parent.__init(self, inputSize, outputSize, rho, false, self.eps, self.momentum, self.affine)
```
Comment (nicholas-leonard) [inline FastLSTM.lua]: you don't need to change line 15 as the last arguments are ignored by the parent.

File: FastLSTM.lua [hunk 2]
```diff
@@ -3,22 +3,46 @@ local FastLSTM, parent = torch.class("nn.FastLSTM", "nn.LSTM")
...
 FastLSTM.usenngraph = false
+FastLSTM.bn = false
 
-function FastLSTM:__init(inputSize, outputSize, rho)
-   parent.__init(self, inputSize, outputSize, rho, false)
+function FastLSTM:__init(inputSize, outputSize, rho, eps, momentum, affine)
+   --  initialize batch norm variance with 0.1
+   self.eps = self.eps or 0.1
+   self.momentum = self.momentum or 0.1 --gamma
+   self.affine = self.affine or true
```
Comment (nicholas-leonard) [inline FastLSTM.lua]:
```lua
self.eps = eps or 0.1
self.momentum = momentum or 0.1 --gamma
self.affine = affine == nil and true or affine
```

File: FastLSTM.lua [hunk 3]
```diff
@@ -3,22 +3,46 @@ local FastLSTM, parent = torch.class("nn.FastLSTM", "nn.LSTM")
...
    -- input : {input, prevOutput, prevCell}
    -- output : {output, cell}
+
+   --apply recurrent batch normalization 
+   -- http://arxiv.org/pdf/1502.03167v3.pdf
+   --normalize recurrent terms W_h*h_{t-1} and W_x*x_t separately 
+
+   --Olalekan Ogunmolu <lexilighty@gmail.com>
+
+   local bn_wx, bn_wh, bn_c
```
Comment (nicholas-leonard) [inline FastLSTM.lua]: Move this directly into the `nngraphModel()` function.

File: FastLSTM.lua [hunk 4]
```diff
@@ -80,7 +104,7 @@ function FastLSTM:buildModel()
    return seq
 end
 
-function FastLSTM:nngraphModel()
+function FastLSTM:nngraphModel(bn_wx, bn_wh, bn_c)
```
Comment (nicholas-leonard) [inline FastLSTM.lua]: By moving the above into `nngraphModel` no need for these function arguments.

File: FastLSTM.lua [hunk 5]
```diff
@@ -91,8 +115,8 @@ function FastLSTM:nngraphModel()
    local x, prev_h, prev_c = unpack(inputs)
    
    -- evaluate the input sums at once for efficiency
-   local i2h = self.i2g(x):annotate{name='i2h'}
-   local h2h = self.o2g(prev_h):annotate{name='h2h'}
+   local i2h = bn_wx(self.i2g(x):annotate{name='i2h'}):annotate {name='bn_wx'}
```
Comment (nicholas-leonard) [inline FastLSTM.lua]:
```lua
if self.bn then
   -- new stuff
else
   -- oldstuff
end
```

**Comment by nicholas-leonard** [comment]
@lakehanne The PR looks good. Could you make the above changes, add a small unit test and some documentation?

File: FastLSTM.lua (hunk 5 above)
Comment (lakehanne) [inline FastLSTM.lua]: We are also batch normalizing the `next_c` after these operations and between batch normalizing `i2h` , `h2h` and `next_c`, we have to do some computations for the gates in-between. If I go with your direction, this would make the code longer. The `nn.Identty()` takes care of the redundancy. Okay anyway if you prefer it to be that way. 

Your implementation would look like 

```lua
   local bn_wx, bn_wh, bn_c  
   if self.bn then  
      bn_wx = nn.BatchNormalization(4*self.outputSize, self.eps, self.momentum, self.affine)
      bn_wh = nn.BatchNormalization(4*self.outputSize, self.eps, self.momentum, self.affine)
      bn_c  = nn.BatchNormalization(self.outputSize, self.eps, self.momentum, self.affine)

      
      -- evaluate the input sums at once for efficiency
      local i2h = bn_wx(self.i2g(x):annotate{name='i2h'}):annotate {name='bn_wx'}
      local h2h = bn_wh(self.o2g(prev_h):annotate{name='h2h'}):annotate {name = 'bn_wh'}
      local all_input_sums = nn.CAddTable()({i2h, h2h})

      local reshaped = nn.Reshape(4, self.outputSize)(all_input_sums)
      -- input, hidden, forget, output
      local n1, n2, n3, n4 = nn.SplitTable(2)(reshaped):split(4)
      local in_gate = nn.Sigmoid()(n1)
      local in_transform = nn.Tanh()(n2)
      local forget_gate = nn.Sigmoid()(n3)
      local out_gate = nn.Sigmoid()(n4)
      
      -- perform the LSTM update
      local next_c           = nn.CAddTable()({
        nn.CMulTable()({forget_gate, prev_c}),
        nn.CMulTable()({in_gate,     in_transform})
      })
      -- gated cells form the output
      local next_h = nn.CMulTable()({out_gate, nn.Tanh()(bn_c(next_c):annotate {name = 'bn_c'}) })
   else
      -- evaluate the input sums at once for efficiency
      local i2h = self.i2g(x):annotate{name='i2h'}
      local h2h = self.o2g(prev_h):annotate{name='h2h'}
      local all_input_sums = nn.CAddTable()({i2h, h2h})

      local reshaped = nn.Reshape(4, self.outputSize)(all_input_sums)
      -- input, hidden, forget, output
      local n1, n2, n3, n4 = nn.SplitTable(2)(reshaped):split(4)
      local in_gate = nn.Sigmoid()(n1)
      local in_transform = nn.Tanh()(n2)
      local forget_gate = nn.Sigmoid()(n3)
      local out_gate = nn.Sigmoid()(n4)
      
      -- perform the LSTM update
      local next_c           = nn.CAddTable()({
        nn.CMulTable()({forget_gate, prev_c}),
        nn.CMulTable()({in_gate,     in_transform})
      })
      -- gated cells form the output
      local next_h = nn.CMulTable()({out_gate, nn.Tanh()(next_c)})
   end
```

You want me to send the PR as suggested?

**Comment by nicholas-leonard** [comment]
>We are also batch normalizing the next_c after these operations and between batch normalizing i2h , h2h and next_c, we have to do some computations for the gates in-between. If I go with your direction, this would make the code longer. The nn.Identty() takes care of the redundancy. Okay anyway if you prefer it to be that way.

>Your implementation would look like 

@lakehanne Good point. Can you separate the `if` into 2 `if`s, one for `bn_wx` and `bn_wh` and the other for `bn_c`. So basically, all the code in between would be written only once, thus minimizing redundancy.

**Comment by lakehanne** [comment]
@nicholas-leonard I made the changes you suggested. Could you take a look at the `readme`, `FastLSTM` class and `recurrent-language-model` example and let me know what you think please?

**Comment by nicholas-leonard** [comment]
@lakehanne I made some additional modifications to your PR (see commit https://github.com/Element-Research/rnn/commit/34e12a2045b1642d568bd9962eabf13c40b97614). Thank you!

**Comment by lakehanne** [comment]
Ice cool vertica ... :100:
